numpy/core/src/multiarray/datetime.c

```diff
@@ -430,6 +431,104 @@ npy_int64 extract_unit(npy_datetime *d, npy_datetime unit) {
     return div;
 }
 
+/*
+ * Convert datetime64 to a tuple (quad_centuries, microseconds, attoseconds)
+ * then hash the tuple.
+ * quad_centuries, because it's easy to convert both (number_of_years) and
+ * (number_of_days) to (number_of_400_year_periods).
+ * microseconds, because 400 years of microseconds fits into 54 bits.
+ * attoseconds, because that's the highest resolution.
+ */
+NPY_NO_EXPORT npy_int64
+datetime_hash(PyArray_DatetimeMetaData *meta, npy_datetime dt)
+{
+    npy_datetimestruct dts;
```

Comment:
Can you just call an existing helper to populate this struct?

Reply:
I could call "convert_datetime_to_datetimestruct", but this would only be used on two of the switch cases ("Y" and "M" units"). And if I used "convert_datetime_to_datetimestruct" then I would have to move the "dt *= meta->num" to avoid doing it twice. It's only 3 lines of code I would be replacing.

Reply:
What I mean is, can you eliminate the entire switch by using `convert_datetime_to_datetimestruct`?

Reply:
Oh I see. But I don't want to convert the datetime to a datetimestruct :-) I don't want a datetimestruct. The calculations don't use (year, month, day). They use seconds, attoseconds etc. Which is quicker and simpler I believe, because it doesn't involve converting to a datetimestruct. It just involves a couple of division operations.

Reply:
Fair enough, looking further we already have a bunch of similar conversions anyway.